src/zjs_promise.c

```diff
@@ -138,29 +117,23 @@ void zjs_fulfill_promise(jerry_value_t obj, jerry_value_t argv[], uint32_t argc)
     }
 
     jerry_get_object_native_handle(promise_obj, (uintptr_t*)&handle);
-
+    jerry_release_value(promise_obj);
     if (handle) {
-        // Put *something* here in case it never gets registered
-        if (!handle->then_set) {
-            jerry_release_value(handle->then);
-            handle->then = jerry_create_external_function(null_function);
-        }
-
-        handle->then_id = zjs_add_callback_once(handle->then,
-                                                handle->this,
+        handle->then_id = zjs_add_callback_once(ZJS_UNDEFINED,
+                                                obj,
                                                 handle,
                                                 post_promise);
```

Comment:
OK, I'm pretty sure this was wrong before and is still wrong now. You can't overwrite then_id in fulfill, because what if we had a true asynchronous promise and then is set and later on the promise is fulfilled.

If I'm right, this means we need some better tests written - one that truly exercises both the lame "synchronous" case and a real asynchronous case, e.g. with a timer.